numpy/ma/core.py

```diff
@@ -5108,77 +5043,143 @@ def anom(self, axis=None, dtype=None):
-                else:
-                    out.flat = np.nan
+                    if any(zerocounts):
+                        out.__setmask__(zerocounts)
+                    else:
+                        out.__setmask__(nomask)
                 return out
-        # In case with have an explicit output
+            # mask in case of empty slices
+            if any(zerocounts):
+                ret = ret.view(MaskedArray)
+                ret.__setmask__(zerocounts)
+            return ret
+
+        # code largely copied from _methods.var
+
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        # Compute the mean, keeping same dims. Note that if dtype is not of
+        # inexact type then arraymean will not be either.
+        rcount = self.count(axis=axis, keepdims=True)
+        arrmean = self.sum(axis=axis, dtype=dtype, keepdims=True)
+
+        # quit now if sum is masked, to avoid writing to the masked singleton
+        if arrmean is masked:
+            return _return_masked(out)
+
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(arrmean, np.ndarray):
+                arrmean = np.true_divide(
+                    arrmean, rcount, out=arrmean, casting='unsafe', subok=False)
+            else:
+                arrmean = arrmean.dtype.type(arrmean / rcount)
+
+        # Compute sum of squared deviations from mean
+        # Note that x may not be inexact and that we need it to be an array,
+        # not a scalar.
+        x = asanyarray(self - arrmean)
+        if issubclass(self.dtype.type, np.complexfloating):
+            x = np.multiply(x, np.conjugate(x), out=x).real
+        else:
+            x = np.multiply(x, x, out=x)
+        ret = x.sum(axis, dtype, **kwargs)
+        if ret is masked:
+            return _return_masked(out)
+
+        # Compute degrees of freedom and make sure it is not negative.
+        rcount = self.count(axis=axis, **kwargs)
+        rcount = (rcount - ddof).clip(min=0)
+
+        # divide by degrees of freedom
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
+            elif hasattr(ret, 'dtype'):
+                ret = ret.dtype.type(ret / rcount)
+            else:
+                ret = ret / rcount
+
+        # account for case we got a scalar from sum, and rcout == 0
+        if ret.ndim == 0 and np.isnan(ret):
+            ret = masked
+
         if out is not None:
-            # Set the data
-            out.flat = dvar
-            # Set the mask if needed
             if isinstance(out, MaskedArray):
-                out.__setmask__(dvar.mask)
+                out.flat = ret
```

Comment:
Why flat? I get the flat assignment for the mask. But here I am not sure. And since flat assignment is buggy as hell....